File: IAIso-v5.0/core/iaiso-node/src/core/execution.ts
/**
 * BoundedExecution — high-level facade combining engine + consent + audit.
 *
 * Node idiomatic usage:
 *
 *   await BoundedExecution.run(
 *     { config: new PressureConfig(), consent, auditSink },
 *     async (exec) => {
 *       while (!done) {
 *         if (exec.check() === "escalated") {
 *           await escalateToHuman();
 *           exec.reset();
 *           continue;
 *         }
 *         exec.requireScope("tools.search");
 *         const result = await runTool("search", query);
 *         exec.recordToolCall({ name: "search" });
 *         exec.recordTokens(result.tokenCount, "search_result");
 *       }
 *     },
 *   );
 *
 * Or explicit lifecycle:
 *   const exec = BoundedExecution.start({ config: new PressureConfig() });
 *   try { ... } finally { exec.close(); }
 */

import { randomUUID } from "node:crypto";

import { AuditEvent } from "../audit/event.js";
import type { AuditSink } from "../audit/sinks/memory.js";
import { NullSink } from "../audit/sinks/memory.js";
import type { ConsentScope } from "../consent/index.js";
import { InsufficientScope } from "../consent/index.js";
import {
  Lifecycle,
  StepOutcome,
} from "./types.js";
import type { Clock } from "./types.js";
import {
  PressureConfig,
  PressureEngine,
  StepInput,
} from "./engine.js";
import type { PressureSnapshot } from "./engine.js";

export class ExecutionLocked extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ExecutionLocked";
  }
}

export class ScopeRequired extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ScopeRequired";
  }
}

export interface BoundedExecutionOptions {
  execution_id?: string;
  config?: PressureConfig;
  consent?: ConsentScope | null;
  audit_sink?: AuditSink;
  clock?: Clock;
  timestampClock?: Clock;
}

export class BoundedExecution {
  readonly engine: PressureEngine;
  consent: ConsentScope | null;
  readonly audit_sink: AuditSink;
  private readonly _timestampClock: Clock;
  private _closed = false;

  private constructor(
    engine: PressureEngine,
    consent: ConsentScope | null,
    audit_sink: AuditSink,
    timestampClock: Clock,
  ) {
    this.engine = engine;
    this.consent = consent;
    this.audit_sink = audit_sink;
    this._timestampClock = timestampClock;
  }

  static start(opts: BoundedExecutionOptions = {}): BoundedExecution {
    const execId = opts.execution_id ?? `exec-${randomUUID()}`;
    const cfg = opts.config ?? new PressureConfig();
    const sink = opts.audit_sink ?? new NullSink();
    const tsClock = opts.timestampClock ?? (() => Date.now() / 1000);
    const engine = new PressureEngine(cfg, {
      execution_id: execId,
      audit_sink: sink,
      clock: opts.clock,
      timestampClock: tsClock,
    });

    const instance = new BoundedExecution(
      engine,
      opts.consent ?? null,
      sink,
      tsClock,
    );
    if (opts.consent) {
      instance._emit("execution.consent_attached", {
        subject: opts.consent.subject,
        scopes: opts.consent.scopes,
        jti: opts.consent.jti,
      });
    }
    return instance;
  }

  /**
   * Run a callback with a BoundedExecution, emitting `execution.closed` on
   * exit (with exception type if the callback throws).
   */
  static async run<T>(
    opts: BoundedExecutionOptions,
    fn: (exec: BoundedExecution) => Promise<T> | T,
  ): Promise<T> {
    const exec = BoundedExecution.start(opts);
    let thrownName: string | null = null;
    try {
      return await fn(exec);
    } catch (err) {
      thrownName =
        err instanceof Error ? err.name : typeof err === "object" ? "Object" : typeof err;
      throw err;
    } finally {
      exec.close(thrownName);
    }
  }

  recordTokens(tokens: number, tag: string | null = null): StepOutcome {
    return this._account(new StepInput({ tokens, tag }));
  }

  recordToolCall(
    params: { name?: string | null; count?: number; tokens?: number } = {},
  ): StepOutcome {
    return this._account(
      new StepInput({
        tokens: params.tokens ?? 0,
        tool_calls: params.count ?? 1,
        tag: params.name ?? null,
      }),
    );
  }

  recordStep(
    params: {
      tokens?: number;
      tool_calls?: number;
      depth?: number;
      tag?: string | null;
    } = {},
  ): StepOutcome {
    return this._account(
      new StepInput({
        tokens: params.tokens ?? 0,
        tool_calls: params.tool_calls ?? 0,
        depth: params.depth ?? 0,
        tag: params.tag ?? null,
      }),
    );
  }

  private _account(work: StepInput): StepOutcome {
    const outcome = this.engine.step(work);
    if (outcome === StepOutcome.Locked) {
      throw new ExecutionLocked(
        `execution ${this.engine.execution_id} is locked; call reset() before continuing`,
      );
    }
    return outcome;
  }

  /** Return the current outcome without advancing. Safe guard for agent-loop tops. */
  check(): StepOutcome {
    const lc = this.engine.lifecycle;
    if (lc === Lifecycle.Locked) return StepOutcome.Locked;
    if (lc === Lifecycle.Escalated) return StepOutcome.Escalated;
    return StepOutcome.OK;
  }

  /**
   * Require the attached consent to grant `scope`. Emits `consent.granted`
   * or `consent.denied` accordingly.
   */
  requireScope(scope: string): void {
    if (this.consent == null) {
      this._emit("consent.missing", { requested: scope });
      throw new ScopeRequired(
        `scope '${scope}' required but no consent attached`,
      );
    }
    try {
      this.consent.require(scope);
    } catch (err) {
      if (err instanceof InsufficientScope) {
        this._emit("consent.denied", {
          requested: scope,
          granted: this.consent.scopes,
          jti: this.consent.jti,
        });
      }
      throw err;
    }
    this._emit("consent.granted", {
      requested: scope,
      jti: this.consent.jti,
    });
  }

  reset(): PressureSnapshot {
    return this.engine.reset();
  }

  snapshot(): PressureSnapshot {
    return this.engine.snapshot();
  }

  close(exception: string | null = null): void {
    if (this._closed) return;
    this._closed = true;
    this._emit("execution.closed", {
      final_pressure: this.engine.pressure,
      final_lifecycle: this.engine.lifecycle,
      exception,
    });
  }

  private _emit(kind: string, data: Record<string, unknown>): void {
    this.audit_sink.emit(
      new AuditEvent(
        this.engine.execution_id,
        kind,
        this._timestampClock(),
        data,
      ),
    );
  }
}

File: IAIso-v5.0/core/iaiso-python/iaiso/core/execution.py
"""BoundedExecution — high-level abstraction combining engine + consent + audit.

This is the primary public API for embedding IAIso into an agent loop. A
BoundedExecution owns a PressureEngine, optionally a ConsentScope, and an
AuditSink. It exposes operations for accounting for work (`record_tokens`,
`record_tool_call`) and for controlling flow (`check`, `require_scope`).

Typical usage:

    with BoundedExecution.start(
        execution_id="exec-abc",
        config=PressureConfig(),
        consent=verified_scope,
        audit_sink=audit_sink,
    ) as exec:
        while not done:
            if exec.check() is StepOutcome.ESCALATED:
                escalate_to_human()
                exec.reset()  # or break, depending on policy
                continue
            exec.require_scope("tools.search")
            result = run_tool("search", query)
            exec.record_tool_call(name="search")
            exec.record_tokens(tokens=result.token_count, tag="search_result")
"""

from __future__ import annotations

import contextlib
import uuid
from dataclasses import dataclass
from types import TracebackType
from typing import Any, Iterator

from iaiso.audit import AuditEvent, AuditSink, NullSink
from iaiso.consent import ConsentScope, InsufficientScope
from iaiso.core.engine import (
    Lifecycle,
    PressureConfig,
    PressureEngine,
    PressureSnapshot,
    StepInput,
    StepOutcome,
)


class ExecutionLocked(RuntimeError):
    """Raised when attempting to record work on a locked execution."""


class ScopeRequired(RuntimeError):
    """Raised when an operation requires a consent scope but none was attached."""


@dataclass
class BoundedExecution:
    """A bounded agent execution with integrated pressure, consent, and audit.

    Do not construct directly; use `BoundedExecution.start()` or `.attach()`.
    """

    engine: PressureEngine
    consent: ConsentScope | None
    audit_sink: AuditSink

    @classmethod
    def start(
        cls,
        *,
        execution_id: str | None = None,
        config: PressureConfig | None = None,
        consent: ConsentScope | None = None,
        audit_sink: AuditSink | None = None,
    ) -> "BoundedExecution":
        """Create a fresh BoundedExecution."""
        exec_id = execution_id or f"exec-{uuid.uuid4()}"
        cfg = config or PressureConfig()
        sink = audit_sink or NullSink()
        engine = PressureEngine(cfg, execution_id=exec_id, audit_sink=sink)

        instance = cls(engine=engine, consent=consent, audit_sink=sink)
        if consent is not None:
            instance._emit("execution.consent_attached",
                           subject=consent.subject,
                           scopes=consent.scopes,
                           jti=consent.jti)
        return instance

    def record_tokens(self, tokens: int, *, tag: str | None = None) -> StepOutcome:
        """Account for generated tokens."""
        return self._account(StepInput(tokens=tokens, tag=tag))

    def record_tool_call(
        self,
        *,
        name: str | None = None,
        count: int = 1,
        tokens: int = 0,
    ) -> StepOutcome:
        """Account for one or more tool calls, plus any tokens they returned."""
        return self._account(
            StepInput(tokens=tokens, tool_calls=count, tag=name)
        )

    def record_step(
        self,
        *,
        tokens: int = 0,
        tool_calls: int = 0,
        depth: int = 0,
        tag: str | None = None,
    ) -> StepOutcome:
        """Account for a generic unit of work."""
        return self._account(StepInput(
            tokens=tokens,
            tool_calls=tool_calls,
            depth=depth,
            tag=tag,
        ))

    def _account(self, work: StepInput) -> StepOutcome:
        outcome = self.engine.step(work)
        if outcome is StepOutcome.LOCKED:
            raise ExecutionLocked(
                f"execution {self.engine.execution_id} is locked; "
                "call reset() before continuing"
            )
        return outcome

    def check(self) -> StepOutcome:
        """Return the current outcome without accounting for new work.

        Useful as a guard at the top of an agent loop to detect escalation
        without double-counting.
        """
        lifecycle = self.engine.lifecycle
        if lifecycle is Lifecycle.LOCKED:
            return StepOutcome.LOCKED
        if lifecycle is Lifecycle.ESCALATED:
            return StepOutcome.ESCALATED
        return StepOutcome.OK

    def require_scope(self, scope: str) -> None:
        """Raise if no consent is attached or the attached consent does not grant `scope`.

        Emits a `consent.check` audit event on every call, successful or not.
        """
        if self.consent is None:
            self._emit("consent.missing", requested=scope)
            raise ScopeRequired(
                f"scope '{scope}' required but no consent attached"
            )
        try:
            self.consent.require(scope)
        except InsufficientScope as exc:
            self._emit("consent.denied",
                       requested=scope,
                       granted=self.consent.scopes,
                       jti=self.consent.jti)
            raise
        self._emit("consent.granted",
                   requested=scope,
                   jti=self.consent.jti)

    def reset(self) -> PressureSnapshot:
        return self.engine.reset()

    def snapshot(self) -> PressureSnapshot:
        return self.engine.snapshot()

    def _emit(self, kind: str, **data: Any) -> None:
        import time
        self.audit_sink.emit(AuditEvent(
            execution_id=self.engine.execution_id,
            kind=kind,
            timestamp=time.time(),
            data=data,
        ))

    def __enter__(self) -> "BoundedExecution":
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        self._emit(
            "execution.closed",
            final_pressure=self.engine.pressure,
            final_lifecycle=self.engine.lifecycle.value,
            exception=type(exc).__name__ if exc else None,
        )
